### DISCUSSION TITLE
LS 2.3.0+git@008f27574: cross: Error 7: Invalid value: That source is fallible.

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3655] [User Support]
**Describe the bug**
Using the LS 2.3.0+git@008f27574 `.deb` on Linux Mint 21.3, I’m experimenting with the crossfades again but keep getting

```
At test-fade2.liq, line 14, char 8-69:
radio = cross(duration=4.00, width=2.00, live_aware_crossfade, radio)

Error 7: Invalid value:
That source is fallible.
```

**To Reproduce (minimal script)**
```
# test-fade2.liq

uri = "~/Musik/Playlists/Radio/Hard Rock.m3u"
radio = mksafe(playlist(uri))

# Be sure to have ReplayGain applied before crossing.
radio = amplify(1.,override="replaygain_track_gain",radio)

def live_aware_crossfade(old, new) =
    log("Using custom crossfade")
    cross.simple(old.source, new.source, fade_in=0.1, fade_out=3.5)
end

radio = cross(duration=4.00, width=2.00, live_aware_crossfade, radio)

output.ao(radio)
```

**Expected behavior**
No error.

**Am I overlooking something here, or is this a bug?**

**Comment by Moonbase59** [comment]
Updated to LS 2.3.0+git@3987e0ac6—same result.

Could this be related to #2080?

**Comment by Moonbase59** [comment]
Same happens when I replace

```
radio = cross(duration=4.00, width=2.00, live_aware_crossfade, radio)
```

by

```
radio = crossfade(radio)
```

**Comment by toots** [comment]
Crossfade is always fallible. This is because the crossfading function can be anything, including something that returns a fallible source.

**Comment by Moonbase59** [comment]
So how to make this work?

Wanted to pre-listen/record my fading tests for later inclusion into AzuraCast. Code is actually a stripped-down version of what’s used in AzuraCast—why does it work there?

**Reply by gAlleb** [reply]
> why does it work there?

It works cause it has fallback to a single file:
```
radio = fallback(id="safe_fallback", track_sensitive = false, [radio, error_file])
```

> So how to make this work?

radio = mksafe(radio) before output:

```
uri = "~/Musik/Playlists/Radio/Hard Rock.m3u"
radio = playlist(uri)
# Be sure to have ReplayGain applied before crossing.
radio = amplify(1.,override="replaygain_track_gain",radio)
def live_aware_crossfade(old, new) =
    log("Using custom crossfade")
    cross.simple(old.source, new.source, fade_in=0.1, fade_out=3.5)
end
radio = cross(duration=4.00, width=2.00, live_aware_crossfade, radio)

radio = mksafe(radio)

output.ao(radio)
```

**Reply by Moonbase59** [reply]
@gAlleb Thanks, you saved my day! :-)

It works but I still have to get my head around this, because I had a `mksafe()` around the playlist and don’t really understand why it produced the error in that case. Non-procedural programming is still partly a mystery for me… ;-)

**Comment by Moonbase59** [comment]
Does anyone know how I can **skip the base part of a song** except the first and last 15 seconds, so I can concentrate on listening to the crossfades instead of having to listen to the whole song?

So for every song, it would play the first 15s, then skip to 15s before the song end, then run into the next crossfade?

**Comment by Moonbase59** [comment]
Okay, think I got that one right:

```
# Skip base part of song, play only first and last 15 seconds
skipped = ref(false)
def skip_song(s) =
  if source.elapsed(s) > 15. then
    if skipped() == false then
      # seeking can take a while!
      s = source.seek(s, source.remaining(s) - 15.)
      log(label="skip_song", "Skipping #{s}s…")
      skipped := true
    end
  else
    skipped := false
  end
end
radio = source.on_frame(radio, {skip_song(radio)})
```

I wonder if this might be handled in a more elegant way, without hitting the streaming loop on every frame… Could be benefical for a "pre-listen" feature (think Amazon), i.e. play 30s of a song, starting at 30s _into_ the song.

**Comment by gAlleb** [comment]
> It works but I still have to get my head around this, because I had a mksafe() around the playlist and don’t really understand why it produced the error in that case.

Well, I suppose that it is due to `crossfade` operator itself. After doing its thing the operator returns a `new source` and it is fallible. First we have source with few tracks (playlist) and we can mksafe it. But after `crossfade` it is a new source with one long track :) So we need to add `mksafe `option or specify a `single security track` for source to fallback.

Moreover crossfade is designed in such way that it needs to change the time at which the source flows therefore it has its own clock.

> Okay, think I got that one right

Does it work with crossfade?


> Could be benefical for a "pre-listen" feature (think Amazon), i.e. play 30s of a song, starting at 30s into the song.

Thought of a way to achieve this with
```
radio = playlist(prefix="annotate:liq_cue_in=30.00,liq_cue_out=60.00", "~/Musik/Playlists/Radio/Hard Rock.m3u")
```

> Does anyone know how I can skip the base part of a song except the first and last 15 seconds, so I can concentrate on listening to the crossfades instead of having to listen to the whole song?

Won't be annotation enough for test purposes? Just "cut out" first 30 seconds and listen to crossfades :)
```
radio = playlist(prefix="annotate:liq_cue_in=0.00,liq_cue_out=30.00", "~/Musik/Playlists/Radio/Hard Rock.m3u")
```

**Reply by Moonbase59** [reply]
> Does it work with crossfade?

Yeah, looks like. If seeking is fast enough, that is. I’m testing locally (well, on NFS mounts actually), so it _is_ fast enough.

> Won't be annotation enough for test purposes?

Thought about that, but wanted to see the real song endings (hard/soft) and the xfades in Audacity.

> Pre-listen

Yup, annotations are good enough for a "pre-listen" feature, that’s right.

What I’m missing in LS is a crossfade feature that says "start new source when old source drops below a level of x dB", to exactly set the transition point. That was a big advantage in SAM.

**Reply by gAlleb** [reply]
> What I’m missing in LS is a crossfade feature that says "start new source when old source drops below a level of x dB", to exactly set the transition point. That was a big advantage in SAM.

Yeah, might be a good feature. The fuction should search the last xdB in the end of track and apply something like `liq_cue_out` respectfully. 

There has been a discussion about it here https://github.com/savonet/liquidsoap/issues/1869

**Reply by Moonbase59** [reply]
Thanks for the pointer. I know Warblefly’s scripts, and preprocessing has its merits (as in ReplayGain, or manually setting cue points), but I feel something real powerful like Liquidsoap should be able to do a radio-type segue "on-the-fly". I mean, many others (like SAM) can do it, so why not LS? :-)

And we _do_ already have real helpful things like `blank.skip`.
